File: services/orchestration/agent/context_manager.py
"""
Token-budget-aware context assembly.

Uses a 4-chars-per-token approximation — accurate enough for English prose
without the overhead of running a real tokenizer at runtime.
"""

from dataclasses import dataclass, field
from typing import Any

CHARS_PER_TOKEN = 4


def estimate_tokens(text: str) -> int:
    return max(1, len(text) // CHARS_PER_TOKEN)


@dataclass
class ContextItem:
    text: str
    label: str = ""
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class PackedContext:
    items: list[ContextItem]
    estimated_tokens: int
    truncated: bool


def pack_into_context(
    items: list[ContextItem],
    budget: int,
    reserve: int = 0,
) -> PackedContext:
    """
    Greedily pack items into a token budget.

    Args:
        items:   Content items to pack (taken in order provided).
        budget:  Total token budget.
        reserve: Tokens reserved for system prompt + expected response.
                 Subtracted from budget before packing.
    """
    effective = max(budget - reserve, 0)
    result: list[ContextItem] = []
    used = 0

    for item in items:
        tokens = estimate_tokens(item.text + item.label)
        if used + tokens > effective:
            return PackedContext(items=result, estimated_tokens=used, truncated=True)
        result.append(item)
        used += tokens

    return PackedContext(items=result, estimated_tokens=used, truncated=False)


def render_context(packed: PackedContext, separator: str = "\n\n---\n\n") -> str:
    """Render packed items into a single prompt string."""
    parts = []
    for item in packed.items:
        if item.label:
            parts.append(f"### {item.label}\n{item.text}")
        else:
            parts.append(item.text)
    return separator.join(parts)


def batch_by_token_budget(
    items: list[ContextItem],
    budget_per_batch: int,
) -> list[list[ContextItem]]:
    """
    Split a large item list into batches where each batch fits within a budget.
    Used in the MapReduce pipeline to handle very large source corpora.
    """
    batches: list[list[ContextItem]] = []
    current: list[ContextItem] = []
    current_tokens = 0

    for item in items:
        tokens = estimate_tokens(item.text + item.label)
        if current_tokens + tokens > budget_per_batch and current:
            batches.append(current)
            current = []
            current_tokens = 0
        current.append(item)
        current_tokens += tokens

    if current:
        batches.append(current)

    return batches

File: app/src/app/lib/agent/contextManager.ts
/**
 * Context Manager — token-budget-aware context assembly.
 *
 * Claude's context window is large (200k tokens for Sonnet/Opus), but for
 * the synthesis pipeline we want predictable, cost-controlled requests.
 * We use a rough 4-chars-per-token estimate which is accurate enough for
 * English prose.  The actual tokenizer would be more precise but adds
 * unnecessary runtime overhead here.
 */

const CHARS_PER_TOKEN = 4;

/** Rough token count estimate for a string. */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / CHARS_PER_TOKEN);
}

export interface ContextItem {
  text: string;
  label?: string; // optional heading shown before the text in the prompt
  metadata?: Record<string, unknown>;
}

export interface PackedContext {
  items: ContextItem[];
  estimatedTokens: number;
  truncated: boolean; // true if some items were dropped to fit the budget
}

/**
 * Greedily pack items into a token budget.
 * Items are taken in the order provided; once the budget is exceeded the
 * remaining items are dropped (set truncated = true).
 *
 * @param items    Content items to pack
 * @param budget   Max tokens to allow
 * @param reserve  Extra tokens to reserve for the system prompt + response
 *                 (subtracted from budget before packing).
 */
export function packIntoContext(
  items: ContextItem[],
  budget: number,
  reserve = 0
): PackedContext {
  const effective = Math.max(budget - reserve, 0);
  const result: ContextItem[] = [];
  let used = 0;

  for (const item of items) {
    const tokens = estimateTokens(item.text + (item.label ?? ""));
    if (used + tokens > effective) {
      return { items: result, estimatedTokens: used, truncated: true };
    }
    result.push(item);
    used += tokens;
  }

  return { items: result, estimatedTokens: used, truncated: false };
}

/**
 * Render a packed context into a single prompt string.
 * Each item is prefixed with its label (if any) as a markdown heading.
 */
export function renderContext(packed: PackedContext, separator = "\n\n---\n\n"): string {
  return packed.items
    .map((item) => (item.label ? `### ${item.label}\n${item.text}` : item.text))
    .join(separator);
}

/**
 * Split a large list of items into batches that each fit within a token budget.
 * Useful for the Map phase when processing many sources in parallel groups.
 */
export function batchByTokenBudget(
  items: ContextItem[],
  budgetPerBatch: number
): ContextItem[][] {
  const batches: ContextItem[][] = [];
  let currentBatch: ContextItem[] = [];
  let currentTokens = 0;

  for (const item of items) {
    const tokens = estimateTokens(item.text + (item.label ?? ""));
    if (currentTokens + tokens > budgetPerBatch && currentBatch.length > 0) {
      batches.push(currentBatch);
      currentBatch = [];
      currentTokens = 0;
    }
    currentBatch.push(item);
    currentTokens += tokens;
  }

  if (currentBatch.length > 0) {
    batches.push(currentBatch);
  }

  return batches;
}
